### ISSUE TITLE
String#-format: Can't embed backslash before interpolation

### ISSUE THREAD

**Issue by aspiers** [supercollider/supercollider#3571]
Here's another situation where a solution to #3570 would be really helpful.  I want to be able to render a symbol as a string immediately preceded by a backslash, e.g.

    (
    var sym = \foo;
    "symbol is \\\\%".format(sym).postln;  // ideally outputs "\foo"
    )

but it seems impossible regardless of how many backslashes I use.

**Comment by brianlheim** [comment]
Oof, this is impossible to fix without breakage. The design is that a single backslash should pass through without modification, but backslash-plus-percent should be replaced with percent. i.e. backslash doesn't need to be escaped despite being the escape char.

You can get this with `"symbol is %%".format("\\", sym)`.

Not sure what to do about this one. We could document it as a known issue; add a new format method with correct behavior; or make the breaking change. I don't know if the last two options are really worth the trouble for this particular case; I think #3570 is a more general solution and I'd prefer to put time into that. What do you think?

**Comment by aspiers** [comment]
I totally agree! Document the workaround and then solve via proper `printf` style approach.

**Comment by telephon** [comment]
In any case it is a little error prone to write it this way.
```supercollider
(
var sym = 'foo bar';
"symbol is %%".format("\\", sym)
)
```

gives you:

`symbol is \foo bar`

which isn't what you wanted.

In such cases I always write: `"symbol is '%'".format(sym)`

That said, it would be kind of nice to have `asCompileString` for symbol return `\something` if it has no spaces.

**Comment by beatboxchad** [comment]
@brianlheim 

> The design is that a single backslash should pass through without modification, but backslash-plus-percent should be replaced with percent. i.e. backslash doesn't need to be escaped despite being the escape char.

Correct me if I'm wrong (weird brain day), but this doesn't seem to match the documentation, which simply states that `\` is the escape character, `%` is the interpolation character, and `\\%` is how you escape the interpolation character (there's a hard-coded check for this in the implementation). When I pass a single backslash without modification, it doesn't show up in the formatted string. It escapes whatever character comes next, even if that character didn't need escaping.

This makes sense given the [implementation](https://github.com/supercollider/supercollider/blob/Version-3.10.3/lang/LangPrimSource/PyrStringPrim.cpp#L153), which is pretty naive. No look-behind, no checking whether a character even needs to be escaped. Only code for the happy case. No wonder it's not behaving intuitively. I know the phrase "undefined behavior" is more typically used for C code that compiles differently depending on the compiler or platform, but I'll invoke it here in arguing that anything more specific and rigorous would be an improvement to the developer experience, even if somebody's code breaks. However, that discussion would be much better suited to an RFC. I happen to be working on one, but I thought I'd info-dump here, because it turns out sclang has a number of competing ideas/implementations around how to format strings.

So I think I thought of something simpler _for this specific issue_. You can get the type of a `Slot` in the sclang primitive system. I think the best solution here, keeping with the simple no-nonsense implementation, is to automatically render slot args which are Symbols in a pretty way without requiring the user to manually escape anything. If there is any whitespace, wrap it in single quotes; otherwise, prepend a `\`. We can document that, too, and nothing else will change.

Deeper consideration around the string formatting facilities in sclang is still very much merited. I haven't even looked at `postf` and friends... I hope they at least use this primitive.

**Comment by telephon** [comment]
The implementation doesn't ever get the original object passed, it is converted to a string first. The place where this would have to happen is in the sclang code of `String::format`:

```supercollider
format { arg ... items; ^this.prFormat( items.collect(_.asString) ) }
```

So your proposal would mean that we want a separate message like `asStringForFormat`, so it would read like:

```supercollider
format { arg ... items; ^this.prFormat( items.collect(_.asStringForFormat) ) }
```

Where `Symbol` responds to it as follows:

```supercollider
+ Symbol {

    asStringForFormat { 
       ^if(this.hasWhiteSpaceOrQuote) { this.asCompileString } { "\\" ++ this }
   }

}
```

Which would need a primitive for the method `hasWhiteSpaceOrQuote`.

**Comment by jamshark70** [comment]
> I think the best solution here, keeping with the simple no-nonsense implementation, is to automatically render slot args which are Symbols in a pretty way without requiring the user to manually escape anything.

We know from the recent experience changing the way Floats print that this *will* have dozens of unforeseen consequences that we will be discovering for months afterward.

I would urge a high degree of caution about this.

**Comment by beatboxchad** [comment]
Thank you both! I have been edified.

Hey @jamshark70 you must be talking about https://github.com/supercollider/supercollider/pull/3585

Yeah... I see... I'm gonna pipe down over here and go back to that RFC. Knowing what I now know, I'm inclined to focus on #3570 and leave everything else alone.
